examples/Coffee_Machine_FSM/coffee_active_learning.py

```diff
@@ -0,0 +1,184 @@
+        return CoffeeObservation(
+            status_code=self.status_code,
+            previous_state=self.previous_state,
+            current_state=self.coffee_machine.get_current_state(),
+        )
+
+# ============================
+# CoffeeModel Class
+# ============================
+class CoffeeModel(Model):
+    def __init__(self):
+        # Storing the best action per state
+        self.best_actions: Dict[str, Action] = {}
+
+    def predict(self, input_features: pl.DataFrame) -> pl.DataFrame:
+        actions = []
+        for state in input_features['current_state']:
+            if state in self.best_actions:
+                actions.append(self.best_actions[state])
+            else:
+                # Default action if state is unknown
+                actions.append('clean')
+        return pl.DataFrame({'action': actions})
+
+    def save(self, path: Path) -> None:
+        pass
+
+    def load(self, path: Path) -> None:
+        pass
+
+# ============================
+# CoffeeLearner Class
+# ============================
+class CoffeeLearner(ActiveLearner[Action, CoffeeObservation]):
+    def __init__(self):
+        self.model = CoffeeModel()
+        self.rewards: List[float] = []
+        self.actions = ['water', 'pod', 'button', 'clean']
+        self.state_action_rewards: Dict[str, Dict[str, float]] = {}
+        self.action_indices: Dict[str, int] = {}
+
+    def learn_active(self, action: Action, observation: CoffeeObservation) -> Model:
+        status_code = observation.status_code
+        previous_state = observation.previous_state
+
+        # Assigning rewards based on status_code
+        if status_code == StatusCode.COFFEE_PRODUCED:
+            reward = 1.0
+            self.model.best_actions[previous_state] = action
+        elif status_code == StatusCode.OK:
+            reward = 0.5
+        elif status_code == StatusCode.ERROR:
+            reward = -1.0
+        else:
+            reward = 0.0  
+
+        self.rewards.append(reward)
+
+        # Updating state-action rewards
+        if previous_state not in self.state_action_rewards:
+            self.state_action_rewards[previous_state] = {action: reward}
+        else:
+            self.state_action_rewards[previous_state][action] = reward
+
+        logger.debug(f"Learned: State: {previous_state}, Action: {action}, Reward: {reward}")
+
+        return self.model
+
+    def propose_action(self, observation: CoffeeObservation) -> Action:
+        current_state = observation.current_state
+
+        if current_state == 'f':
+            return 'clean'  
+
+        if current_state in self.model.best_actions:
+            return self.model.best_actions[current_state]
+        else:
+            if current_state not in self.action_indices:
+                self.action_indices[current_state] = 0
+
+            index = self.action_indices[current_state]
+            while index < len(self.actions):
+                action = self.actions[index]
```

Comment:
I don't understand why / what we are doing here. Why don't we select the action with the highest reward and / or sample among the actions depending on their reward..?